### DISCUSSION TITLE
Superset Group By Filter

### DISCUSSION THREAD

**Discussion by swaza** [apache/superset#15272] [General]
I have a table named fact_sales_revenue_monthly with the following columns sales_month, area, district and revenue. The requirement is to build a dashboard where I have to provide the user a filter box that displays the names of the dimension columns namely  area and district from this table. The user can pick only one dimension name at a time from the filter (area or district)..Based on what dimension column name is chosen , It will display a table chart of revenue by month and value of the dimension column. So if the user select ‘area’ , then it will display the revenue by month for different areas (eg US-WEST, US-EAST) and so on and if the user selects a district then the chart will display revenue for different districts like NORCAL01,NORCAL02 and so on.

I created 2 charts . one chart is of type filter box where displays the user the name of the dimensions namely area and district. The names match with the column names in the Table.
I named the filter column as group_by_2. Next I create the Table chart to display the revenues and connect the filter box to the Table chart.
The Query for the Table chart dataset  looks like below . 

```
    {% set filter = filter_values('group_by_2', 'area')[0] %}
   SELECT
       TRUNC(sales_month::date) AS sales_month,
       {{filter}} AS group_by_2,
       revenue
   FROM fact_sales_revenue_monthly
```

The above  query provides me results with no issues for both area and district.  However When I add the filter box and the table chart on the dashboard and change the filter to district or area it does replace the group_by_2 .However It also adds an additional WHERE CLAUSE  so the resulting query shows as below. Since this a filter It is evident that Superset automatically adds a where clause .group_by_2 =’area’ that obviously gets no results since group_by_2=’area’ is a false condition. If i remove the WHERE clause then it works as expected

```
SELECT sales_month AS sales_month,
       group_by_2 AS group_by_2,
       sum(revenue) AS "revenue"
FROM
  (SELECT TRUNC(sales_month::date) AS sales_month,
          area AS group_by_2,
          revenue
   FROM fact_sales_revenue_monthly) AS expr_qry
WHERE group_by_2 = 'area'
  AND ((to_char(sales_month, 'YYYY') = date_part_year(CURRENT_DATE)))
GROUP BY sales_month,
         group_by_2
LIMIT 10000;
```
How do I get around this issue ? Any help much appreciated

**Comment by kamalkeshavani-aiinside** [comment]
You can try this out.
The query used to create filter chart:
```
SELECT 'area' as dimension
UNION ALL
SELECT 'district' as dimension

```

Query used for chart showing the metric:
```
select *, 
{% if filter_values('dimension')|length %}
 {{ filter_values('dimension')[0] }} 
{% else %}
''
{% endif %}
as grouping_dim
from fact_sales_revenue_monthly
```

You need to keep the column names different in both query, so that when Filter is applied from dashboard, it will not add a WHERE clause in the query.

**Reply by amitagraibm** [reply]
Hi I am not able to get value from dashboard. I tried this option {% if filter_values('dimension')|length %}
 {{ filter_values('dimension')[0] }} 
{% else %}
''
{% endif %} but getting blank value. I need to pass a single date.

**Comment by crosstowntraffic** [comment]
I"ve just started with Superset. And have spent 2 days experimenting so far.
I'm finding that for me the Getting Started documents are lacking the deep drill-down to setting up the queries themselves for different chart types.  I've examined different chart examples in the code that I can access but a deeper tutorial in the beginning start up documentation is what I need to understand the hands-on process.

I keep reading the articles for all these dashboards, charts, business analysis solutions that skip over the deeper preparation of queries that will actually be needed to get comprehensible charts. So many docs are pretending that the chart logic itself will "know what to do if you don't".

What I have to really learn is how to write the SQL queries to whittle down the data that I need for Filters.
My experiments with the SQL editor and how it interprets common queries that I write specifically for mysql tables haven't gotten the results I'd hoped for.

So before I keep experimenting may I just lay out my situation as clearly as I can without a video or screenshots?

I have under a thousand records from an online form fills in 18 table columns per record.
I want to duplicate a chart I"ve seen online with each column named in the table represented in a horizontal stacked chart.
Instead of standing vertically, each column stacked on top of each other.

All columns ending at the same Total record value on the right side of the chart by whatever scale I decide.

 Filters will be based on queries pinpointing the distinct variables contained in 7 column categories.
One filter will be named **Altersgruppe** (Age Group by 4 Categories) in a left vertical column.
Under the Filter name are checkboxes for each filter variables available for that column.

Another Filter name would be **Story Ort** ( 6 or 7 normalized Business locations)
And There will be 5 more Filters named by their table column names 

When the variables are clicked in any order and from the full array of filter variables the horizonal columns begin to change length and totals.

In Apache Superset I’ve tried different ways to write queries to isolate each of the filter variables that will display in the Dashboard of my horizontal column stacked chart.

Now I read that I must write these filters in json separately and put that code into the Dashboard itself where I have exported my chart to.

But I don”t understand how Superset applies filters that will be controlled by selecting and unselecting checkboxes or radio buttons.

After not having success with using Conditionals based on "WHERE this … "  (Superset query writing wizard is not completely logical here)

I have tried to use CASE to create filters.

But I don’t understand the internal querying happening in Superset.

**Do I write Case queries for just one column name and its variables and call it by a naming structure in the json Dashboard code?**

**Do I write 7 separate CASE queries defining all the variables that this column field might contain?**

So that when one of the Named filters — such as “Baby-Boomer” is selected in combination with “Electroniks” from another column filter that the entire chart simply recalculates the starting “all records” query with the filters applied or excluded?

Case for column variables in column name 'altersgruppe'

SELECT count(*), 
(case WHEN altersgruppe LIKE 'Baby-Boomer' then 'Baby-Boomer' 
 WHEN altersgruppe LIKE 'Gen X' then 'Gen X' 
 WHEN altersgruppe LIKE 'Gen Y' then 'Gen Y' 
 WHEN altersgruppe LIKE 'Gen Z' then 'Gen Z' 
 else altersgruppe END) AS 'Altersgruppe' 
 FROM story_collection_new 
 GROUP BY altersgruppe

Case for column variables in column name 'story ort' --
SELECT COUNT(*), 
(case 
WHEN
story_ort  LIKE '%Apotheke%' THEN  'Apotheke'
WHEN  story_ort LIKE '%Boutique / Schuhgeschäft%' THEN 'Boutique / Schuhgeschäft'
WHEN story_ort LIKE '%Buchhandel%' THEN 'Buchhandel'
WHEN story_ort LIKE '%Elektronik-Fachgeschäft%'  THEN 'Elektroniks'
WHEN  story_ort LIKE '%Optiker, Akustiker%' THEN 'Optiker, Akustiker'
WHEN  story_ort LIKE '%Sonstige%' THEN 'Sonstige'
else 
 story_ort  END) AS 'Story_Ort'
 FROM story_collection_new
 GROUP BY story_ort

May I please have a demonstration of where the Filter queries are written, how the corresponding json is written that refers to the filter queries?
And how to have more than 2 filter categories that will display in the dashboard using a dropdown list or sub-category list?

& Are the queries best stored into the mysql procedures as virtual tables?

Thank you for your taking the time to bring me up to speed.

**Comment by anshjain18** [comment]
Hi @kamalkeshavani-aiinside , a little late to the party.
Have a basic doubt about the approach discussed here: https://github.com/apache/superset/discussions/15272#discussioncomment-896182
If we keep column names as different, the filters would work fine, but the chart would also display a warning sign that the filter is Incompatible even though the values are correctly displayed.
I understand this is because the chart or the dataset doesn't actually have a column with that name.
Can we do something of that warning sign?
